src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -2335,43 +2380,334 @@ public sealed override void Filter(
+                                testsToAssemble.Push(seq);
+                                testsToRewrite.Push(null); // marker to indicate we need to reassemble after handling children
+                                testsToRewrite.AddRange(seq.RemainingTests!);
+                                break;
+
+                            case null:
+                                var toAssemble = testsToAssemble.Pop();
+                                var length = toAssemble.RemainingTests.Length;
+                                var newSequence = ArrayBuilder<Tests>.GetInstance(length);
+                                for (int i = 0; i < length; i++)
+                                {
+                                    newSequence.Add(testsRewritten.Pop());
+                                }
+
+                                testsRewritten.Push(toAssemble.Update(newSequence));
+                                break;
+
+                            default:
+                                testsRewritten.Push(current.RewriteNestedLengthTests());
+                                break;
+                        }
+                    }
+                    while (testsToRewrite.Count != 0);
+
+                    var result = testsRewritten.Pop();
+
+                    if (!testsRewritten.IsEmpty || !testsToAssemble.IsEmpty)
+                    {
+                        throw ExceptionUtilities.Unreachable();
+                    }
+
+                    testsToRewrite.Free();
+                    testsToAssemble.Free();
+                    testsRewritten.Free();
+
+                    return result;
+                }
+
+                public sealed override bool Equals(object? obj)
+                {
+                    bool? easyOut = equalsEasyOut(this, obj);
+                    if (easyOut.HasValue)
+                    {
+                        return easyOut.GetValueOrDefault();
+                    }
+
+                    Debug.Assert(obj is SequenceTests);
+
+                    var tests1 = ArrayBuilder<Tests>.GetInstance();
+                    var tests2 = ArrayBuilder<Tests>.GetInstance();
+
+                    tests1.AddRange(this.RemainingTests);
+                    tests2.AddRange(((SequenceTests)obj).RemainingTests);
+
+                    do
+                    {
+                        var t1 = tests1.Pop();
+                        var t2 = tests2.Pop();
+
+                        if (t1 is SequenceTests sequence)
+                        {
+                            easyOut = equalsEasyOut(sequence, t2);
+                            if (easyOut.HasValue)
+                            {
+                                if (easyOut.GetValueOrDefault())
+                                {
+                                    continue;
+                                }
+
+                                return false;
+                            }
+
+                            Debug.Assert(t2 is SequenceTests);
```

Comment:
Nit: it's not obvious at first read that `equalsEasyOut` will validate that `RemainingTests` are equal in length. Think we can strengthen this assert and make it easier to understand by asserting that the remaining tests lengths are equal here.